apollo-federation/src/operation/rebase.rs

```diff
@@ -398,6 +398,12 @@ impl InlineFragment {
                 CompositeTypeDefinitionPosition::try_from(rebased_condition_position)
             })
         {
+            // Root types can always be rebased and the type condition is unnecessary.
+            // Moreover, the actual subgraph might have renamed the root types, but the
+            // supergraph schema does not contain that information.
+            if parent_schema.is_root_type(rebased_condition.type_name()) {
```

Comment:
Note this is only safe if the rebased type condition equals the parent type, which you should check for here.

If it doesn't equal the parent type, for now, we can just continue with the rest of the logic. (This means the PR only fixes the bug in some cases, but fixing the bug in those other cases is non-trivial, and those other cases are pretty rare to begin with. The full fix probably won't happen until after the source-awareness initiative is finished.)

Reply:
Thanks. I added a comment about that.

Reply:
Oh, I forgot to push the commit. The check is in place now.